src/relay/pass/unifier.h

```diff
@@ -0,0 +1,132 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file include/tvm/relay/pass/unifier.h
+ * \brief The type unifier which solves a system of equations between
+ * incomplete types.
+ */
+#ifndef TVM_RELAY_PASS_UNIFIER_H_
+#define TVM_RELAY_PASS_UNIFIER_H_
+
+#include <string>
+#include <tvm/relay/expr.h>
+#include "./type_functor.h"
+
+namespace tvm {
+namespace relay {
+
+struct UnionFindError : dmlc::Error {
+  explicit UnionFindError(const std::string& msg) : Error(msg) {}
+};
+
+struct UnificationError : dmlc::Error {
+  explicit UnificationError(const std::string& msg) : Error(msg) {}
+};
+
+struct SubstitutionError : dmlc::Error {
+  explicit SubstitutionError(const std::string& msg) : Error(msg) {}
+};
+
+/*! \brief a union-find data structure for the type-checker */
+class UnionFind;  // forward declaration
+
+class UnionFindNode : public Node {
+ public:
+  tvm::Map<IncompleteType, Type> uf_map;
+
+  UnionFindNode() {}
+
+  void VisitAttrs(tvm::AttrVisitor* v) final { v->Visit("uf_map", &uf_map); }
+
+  TVM_DLL static UnionFind make(tvm::Map<IncompleteType, Type> uf_map);
+
+  // insert v into UF
+  void insert(const IncompleteType& v);
+
+  // infers that v1 and v2 must be of the smae type
+  void unify(const IncompleteType& v1, const Type& v2);
+
+  // returns representative of v's UF-group
+  Type find(const IncompleteType& v);
+
+  void debug();
+
+  void AssertAlphaEqual(const Type& l, const Type& r);
+
+  static constexpr const char* _type_key = "relay.UnionFind";
+  TVM_DECLARE_NODE_TYPE_INFO(UnionFindNode, Node);
+};
+
+class UnionFind : public NodeRef {
+ public:
+  UnionFind() {}
+  explicit UnionFind(std::shared_ptr<tvm::Node> p) : NodeRef(p) {}
+
+  // no const so that union find can be mutable as a member of unifier
+  inline UnionFindNode* operator->() const {
```

Comment:
I think there is no need to put inline before member functions, compiler will automatically inline them if it is necessary/profitable.